File: 1036-rotting-oranges/solution.py
class Solution:
    from collections import deque
    def orangesRotting(self, grid: List[List[int]]) -> int:
        if not grid: return -1
        res = 0
        freshcount = 0
        rowlen, collen = len(grid), len(grid[0])
        q = deque()
        directions = ((0, 1), (0, -1), (-1, 0), (1, 0))
        for row in range(rowlen):
            for col in range(collen):
                if grid[row][col] == 2:
                    q.append((row, col))
                if grid[row][col] == 1:
                    freshcount += 1
        if freshcount == 0:
            return 0
        while q:
            res += 1
            for _ in range(len(q)):
                r, c = q.popleft()
                for r1, c1 in directions:
                    nr, nc = r1 + r, c1 + c
                    if nr < 0 or nc < 0 or nr >= rowlen or nc >= collen:
                        continue
                    if grid[nr][nc] == 1:
                        freshcount -= 1
                        grid[nr][nc] = 2
                        q.append((nr, nc))
            if freshcount == 0:
                return res
        if freshcount > 0:
            return -1
        return res

File: 1036-rotting-oranges/solution.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        deque<pair<int, int>> q;
        int ROWS = grid.size();
        int COLS = grid[0].size();
        for(int r = 0; r < ROWS; ++r){
            for(int c = 0; c < COLS; ++c){
                if(grid[r][c] == 2){
                    q.push_back({r, c});
                }
            }
        }
        int time = 0;
        vector<pair<int, int>> directions = {{1, 0}, {-1, 0}, {0, -1}, {0, 1}};
        while(!q.empty()){
            int size = q.size();
            for(int i = 0; i < size; ++i){
                auto cur = q.front();
                int r = cur.first;
                int c = cur.second;
                q.pop_front();
                grid[r][c] = 2;
                for(auto dire : directions){
                    int nr = r + dire.first;
                    int nc = c + dire.second;
                    if(0 <= nr and 0 <= nc and nr < ROWS and nc < COLS and grid[nr][nc] == 1){
                        grid[nr][nc] = 2;
                        q.push_back({nr, nc});
                    }
                }

            }
            if(!q.empty()){
                time += 1;
            }
        }

        for(int r = 0; r < ROWS; ++r){
            for(int c = 0; c < COLS; ++c){
                if(grid[r][c] == 1){
                    return -1;
                }
            }
        }
        return time;
    }
};
